### DISCUSSION TITLE
How to read a molecule with properties from a string representing a fragment of an SD file?

### DISCUSSION THREAD

**Discussion by drassokhin** [rdkit/rdkit#5747] [Q&A] [answered]
In certain scenarios, it is very convenient to read molecules from an SD source (file, database, standard input) without using Chem.ForwardSDMolSupplier or SDMolSupplier, by first extracting molecular records from the input source as strings one by one, and then converting them to molecules. Chem.MolFromMolBlock can be used for that purpose; however, this method parses just the ctab part of the input string, but ignores the Data fields that follow the ctab block after the M END marker (ForwardSDMolSupplier or SDMolSupplier do read those blocks and convert the key/value parts to the molecule properties), for example,
```
***CTAB part***
M  END
> <id>
123

> <name>
active analog

$$$$
```
Is there a way to parse an SD record from string into a molecule w/properties without first converting just the ctab part with MolFromMolBlock, and then extracting those field names/values from the SD string via text matching and assigning the properties to the molecule with mol.SetProperty(field, value)? Creating a string stream with Python StringIO for each string and then reading just one molecule at a time from that stream with ForwardSDMolSupplier and then the iterator seems like an overkill (I have not done any tests, but my gut feeling is that this is going to waste a lot of CPU cycles).

**Comment by ptosco** [comment, marked as answer]
@drassokhin You may instantiate a `Chem.SDMolSupplier` passing no parameters with the default constructor, and then call its `SetData()` method:
http://rdkit.org/docs/source/rdkit.Chem.rdmolfiles.html?highlight=sdmolsupplier#rdkit.Chem.rdmolfiles.SDMolSupplier.SetData
passing your SD string as data.
Finally, read the SD string into a molecule calling `next()` on the supplier; this will also parse the SD properties.

**Reply by drassokhin** [reply]
@ptosco Hi Paolo! Thanks, you have been very helpful. Yes, the method you suggested does work. It is just a tiny bit slower than using the MolFromMolBlock, but more convenient, because it eliminates the need to manually parse the properties block and add them to the constructed molecule objects. Also, it is not reflected in the documentation that the source parameter to SDMolSupplier constructor (actually, the corresponding ```__init__``` method in the Python wrapper class) is, in fact, optional (the Python linter complains about it). Instead of calling next() on the supplier, one can also use the indexer (```m = ms[0]```), but, to my surprise, the ```next(ms)``` method turned out to be a bit faster.  Here's a working example just for reference:
```Python
def test():
    molstring = '''CHEMBL153534
     RDKit          2D

 16 17  0  0  0  0  0  0  0  0999 V2000
    7.6140  -22.2702    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    5.7047  -23.1991    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    6.1806  -22.5282    0.0000 N   0  0  0  0  0  0  0  0  0  0  0  0
    6.9604  -22.7690    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    4.8790  -23.2163    0.0000 N   0  0  0  0  0  0  0  0  0  0  0  0
    8.2791  -21.1119    0.0000 N   0  0  0  0  0  0  0  0  0  0  0  0
    7.5280  -21.4445    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    8.4225  -22.4364    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    8.8353  -21.7198    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    6.2035  -23.8527    0.0000 S   0  0  0  0  0  0  0  0  0  0  0  0
    4.0534  -23.2163    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    6.9776  -23.5889    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    3.6406  -22.4938    0.0000 N   0  0  0  0  0  0  0  0  0  0  0  0
    3.6406  -23.9215    0.0000 N   0  0  0  0  0  0  0  0  0  0  0  0
    8.4397  -20.3035    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    9.6552  -21.6280    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
  2  3  2  0
  3  4  1  0
  4  1  1  0
  5  2  1  0
  6  7  1  0
  7  1  2  0
  8  1  1  0
  9  8  2  0
 10 12  1  0
 11  5  2  3
 12  4  2  0
 13 11  1  0
 14 11  1  0
 15  6  1  0
 16  9  1  0
  6  9  1  0
 10  2  1  0
M  END
> <chembl_id>
CHEMBL153534

$$$$
'''
    # Can ignore linter missing arg ("parameter p_object unfilled") error here
    ms = Chem.SDMolSupplier()
    try:
        ms.SetData(molstring)
        m = next(ms)
        # This will also work, but is slightly slower
        # m = ms[0]
    # If this code is made part of some reusable library,
    # it would make sense to translate
    # potential exceptions into something more meaningful
    except Exception:
        raise Exception('Error parsing molstring')
    print(m.GetPropsAsDict())
```

**Comment by mgarard** [comment]
Hi,

I don't understand your aversion to BytesIO or StringIO.  It's the same as file manipulation.   You need to convert it one way or another to the proper format set up by rdkit if that's what you're using.  I doubt it would be much different.  I'll provide the examples.

If you're pulling the sdf from a DB as text, here is an example of how you can convert it to an in-memory file object and interpret it with rdkit.

```
molstring = <sdf as text>

from io import BytesIO
mols = Chem.ForwardSDMolSupplier( BytesIO(molstring.encode() ), strictParsing=False )
mol = [ m for m in mols ][0]
print( Chem.MolToSmiles( mol ) )
```

Edit:
A StringIO example is given in the docs for rdkit.
If you want to save SD text back to a DB, here is an example conversion:
```
molio = StringIO()
with Chem.SDWriter( molio ) as w:
  w.write( mol )

sd_text = molio.getvalue()
print( sd_text )
```
